runtime/src/bank.rs

```diff
@@ -4162,7 +4162,7 @@ impl Bank {
                 programs_modified_by_tx,
             } = execution_result
             {
-                if details.status.is_ok() {
+                if details.status.is_ok() && !programs_modified_by_tx.is_empty() {
                     let mut cache = self.transaction_processor.program_cache.write().unwrap();
```

Comment:
actually, i noticed that this is per-tx write-lock while writing this: https://github.com/anza-xyz/agave/pull/1037#discussion_r1579449836

Reply:
You are right. How about:

```Rust
if execution_results.iter().any(|execution_result| matches!(execution_result, TransactionExecutionResult::Executed { details, programs_modified_by_tx } if details.status.is_ok() && !programs_modified_by_tx.is_empty()) {
    let mut cache = self.transaction_processor.program_cache.write().unwrap();
    for execution_result in &execution_results {
        if let TransactionExecutionResult::Executed { programs_modified_by_tx, .. } = execution_result {
            cache.merge(programs_modified_by_tx);
        }
    }
}
```